posts/zzz_DO_NOT_EDIT_document_yo.../document_your_universe_with_dverse.qmd

```diff
@@ -0,0 +1,177 @@
+description: "How to create a global search across an entire collection of R packages: A use case for the toy 'admiralverse'"
+# Note that the date below will be auto-updated when the post is merged.
+date: "2024-12-15"
+# Please do not use any non-default categories.
+# You can find the default categories in the repository README.md
+categories: [Technical]
+# Feel free to change the image
+image: "pharmaverse.png"
+
+---
+
+<!--------------- typical setup ----------------->
+
+```{r setup, include=FALSE}
+long_slug <- "zzz_DO_NOT_EDIT_document_yo..."
+library(link)
+link::auto()
+
+knitr::opts_chunk$set(
+  message = FALSE,
+  collapse = TRUE
+)
+```
+
+<!--------------- post begins here ----------------->
+
+This article shows how to create a global search for any collection of R
+packages, i.e. a universe. Your universe may be on CRAN, large, and popular like
+the [tidyverse](https://tidyverse.org/) or
+[pharmaverse](https://pharmaverse.org/). But it may also be not on CRAN, small,
+company-specific or personal.
+
+The example here uses the 'admiralverse' -- a toy universe defined as every
+package on CRAN today (2024-12-07), with a website, and a name starting with
+'admiral'.
+
+You need:
+
+* the {dverse} package, 
+* each package in the toy 'admiralverse',
+* {dplyr} to manipulate the output of {dverse}, and 
+* {DT} to create a searchable table.
+
+```{r packages}
+library(dverse)
+library(admiral)
+library(admiraldev)
+library(admiralonco)
+library(admiralophtha)
+library(admiralpeds)
+library(admiralvaccine)
+library(dplyr)
+library(DT)
+```
+
+<details><summary>Versions</summary>
+
+```{r versions}
+packageVersion("dverse")
+packageVersion("admiral")
+packageVersion("admiraldev")
+packageVersion("admiralonco")
+packageVersion("admiralophtha")
+packageVersion("admiralpeds")
+packageVersion("admiralvaccine")
+packageVersion("dplyr")
+packageVersion("DT")
+```
+
+</details>
+
+dverse::document_universe() takes the names of the packages in the
+'admiralverse', and a template pointing to each help file (topic) in each
+package.
+
+```{r docs}
+admiralverse <- c(
+  "admiral",
+  "admiraldev",
+  "admiralonco",
+  "admiralophtha",
+  "admiralpeds",
+  "admiralvaccine"
+)
+# For example: https://pharmaverse.github.io/admiral/reference/queries.html
+template <- "https://pharmaverse.github.io/{package}/reference/{topic}.html"
+
+docs <- dverse::document_universe(admiralverse, template)
+docs
+```
+
+You may need to exclude some topics, such as reexported objects from other
+packages. Because topics outside your universe yield broken links, you can
+exclude them with dverse::is_online().
+
+```{r online}
+docs |> filter(!dverse::is_online(topic))
+
+online <- docs |> filter(dverse::is_online(topic))
+```
+
+Consider the different kinds of documentation in your universe.
+
+```{r kinds}
+online |> count(type, keyword)
+```
+
+Here it makes sense to separate vignettes, datasets, and public functions.
+
+### Vignettes
+
+```{r vignettes}
```

Comment:
hmmm - when I render the post locally, the `online` object has no vignettes in it so this table doesn't render nicely. but i can see from your own render that you have some vignettes there. wonder what's up?

![image](https://github.com/user-attachments/assets/3ad4a1a6-5f37-42a4-9cca-22e1c631ea77)

Reply:
Thanks Edoardo! 

Likely you installed the packages in “developer mode”. 

I can reproduce the problem by installing but not building vignettes, which
is the default of `devtools.install()` and `remotes::install_github`
(also of `pak::pak("owner/repo")` but lacks a `build_vignettes` argument).

``` r
formals(devtools::install)["build_vignettes"]
#> $build_vignettes
#> [1] FALSE
formals(remotes::install_github)["build_vignettes"]
#> $build_vignettes
#> [1] FALSE


# WARNING: Runing this reprex will modify your system!


remotes::install_github("pharmaverse/admiral")
#> Using GitHub PAT from the git credential store.
#> Downloading GitHub repo pharmaverse/admiral@HEAD
#> 
#> ── R CMD build ─────────────────────────────────────────────────────────────────
#> * checking for file ‘/tmp/RtmpSNZmQ5/remotes220b42b550823/pharmaverse-admiral-6351451/DESCRIPTION’ ... OK
#> * preparing ‘admiral’:
#> * checking DESCRIPTION meta-information ... OK
#> * checking for LF line-endings in source and make files and shell scripts
#> * checking for empty or unneeded directories
#> * building ‘admiral_1.1.1.9036.tar.gz’
#> Installing package into '/home/rstudio/R/x86_64-pc-linux-gnu-library/4.4'
#> (as 'lib' is unspecified)
library(admiral)
out <- dverse::document_universe("admiral")
# How many vignettes?
sum(out$type == "vignette")
#> [1] 0

remove.packages("admiral")
#> Removing package from '/home/rstudio/R/x86_64-pc-linux-gnu-library/4.4'
#> (as 'lib' is unspecified)

remotes::install_github("pharmaverse/admiral", build_vignettes = TRUE)
#> Using GitHub PAT from the git credential store.
#> Downloading GitHub repo pharmaverse/admiral@HEAD
#> 
#> ── R CMD build ─────────────────────────────────────────────────────────────────
#> * checking for file ‘/tmp/RtmpSNZmQ5/remotes220b4534d11c4/pharmaverse-admiral-6351451/DESCRIPTION’ ... OK
#> * preparing ‘admiral’:
#> * checking DESCRIPTION meta-information ... OK
#> * installing the package to build vignettes
#> * creating vignettes ... OK
#> * checking for LF line-endings in source and make files and shell scripts
#> * checking for empty or unneeded directories
#> * building ‘admiral_1.1.1.9036.tar.gz’
#> Installing package into '/home/rstudio/R/x86_64-pc-linux-gnu-library/4.4'
#> (as 'lib' is unspecified)
#> Adding 'admiral_1.1.1.9036_R_x86_64-pc-linux-gnu.tar.gz' to the cache
library(admiral)
out <- dverse::document_universe("admiral")
# How many vignettes?
sum(out$type == "vignette")
#> [1] 15

remove.packages("admiral")
#> Removing package from '/home/rstudio/R/x86_64-pc-linux-gnu-library/4.4'
#> (as 'lib' is unspecified)

install.packages("admiral")
#> Installing package into '/home/rstudio/R/x86_64-pc-linux-gnu-library/4.4'
#> (as 'lib' is unspecified)
library(admiral)
out <- dverse::document_universe("admiral")
# How many vignettes?
sum(out$type == "vignette")
#> [1] 15
```

<sup>Created on 2024-12-09 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

In conclusion, installing from CRAN with `install.packages()` (or `pak::pak()` should build the vignettes, but the target audience of this post is a developer so it's likely they will also lack the vignettes. 

How about writing a footnote linking to this comment?